x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleAction.java

```diff
@@ -138,32 +167,21 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
         // Mark source index as read-only
         ReadOnlyStep readOnlyStep = new ReadOnlyStep(readOnlyKey, cleanupDownsampleIndexKey, client);
```

Comment:
From the readonly step we can no jump straight to the `DownsampleStep` denoted by `downsampleKey` (there's no need to go through the no-op steps).

The noop steps are there just so deployments that are in that step (that we now converted to a noop) when the Elasticsearch upgrade was performed resume the ILM execution and complete the downsample action after upgrade. (perhaps a comment stating the above would help)